src/services/identity/UsersService.ts

```diff
@@ -188,26 +210,106 @@ class UsersService {
+    const html = `<p>Your OTP is <b>${otp}</b>. It will expire in ${milliSecondsToMinutes(
+      PARSED_EXPIRY
+    )} minutes. Please use this to verify your email address.</p>
     <p>If your OTP does not work, please request for a new OTP.</p>
     <p>IsomerCMS Support Team</p>`
     await this.mailer.sendMail(email, subject, html)
   }
 
   async sendSmsOtp(mobileNumber: string) {
-    const otp = this.otp.generate(mobileNumber)
-    const expiry = this.otp.getExpiryMinutes()
+    const { otp, hashedOtp } = await this.otpService.generateLoginOtpWithHash()
 
-    const message = `Your OTP is ${otp}. It will expire in ${expiry} minutes. Please use this to verify your mobile number`
+    // Reset attempts to login
+    const otpEntry = await this.otpRepository.findOne({
+      where: { mobileNumber },
+    })
+    if (!otpEntry) {
+      await this.createOtpEntry(mobileNumber, OtpType.Mobile, hashedOtp)
+    } else {
+      await otpEntry?.update({ hashedOtp, attempts: 0 })
+    }
+
+    const message = `Your OTP is ${otp}. It will expire in ${milliSecondsToMinutes(
+      PARSED_EXPIRY
+    )} minutes. Please use this to verify your mobile number`
     await this.smsClient.sendSms(mobileNumber, message)
   }
 
-  verifyOtp(value: string, otp: string) {
-    return this.otp.verify(value, otp)
+  async verifyOtp(key: string, keyType: OtpType, otp: string) {
+    let otpEntry
+    if (keyType === OtpType.Email) {
+      otpEntry = await this.otpRepository.findOne({ where: { email: key } })
+    } else {
+      otpEntry = await this.otpRepository.findOne({
+        where: { mobileNumber: key },
+      })
+    }
+
+    if (!otpEntry?.hashedOtp) {
+      // TODO: Change to use AuthError after FE fix
+      throw new BadRequestError("Hashed OTP not found")
+    }
+
+    if (
+      otpEntry.attempts !== null &&
+      otpEntry.attempts >= MAX_NUM_OTP_ATTEMPTS
+    ) {
+      // TODO: Change to use AuthError after FE fix
+      throw new BadRequestError("Max number of attempts reached")
+    }
+
+    // increment attempts
+    await otpEntry.update({ attempts: otpEntry.attempts + 1 })
+
+    const isValidOtp = await this.otpService.verifyOtp(otp, otpEntry.hashedOtp)
+    if (!isValidOtp) {
+      // TODO: Change to use AuthError after FE fix
+      throw new BadRequestError("OTP is not valid")
+    }
+
+    return true
+  }
+
+  private getOtpExpiry() {
+    return new Date(Date.now() + PARSED_EXPIRY)
+  }
+
+  private async createOtpEntry(
+    key: string,
+    keyType: OtpType,
+    hashedOtp: string
+  ) {
+    if (keyType === OtpType.Email) {
+      await this.otpRepository.create({
+        email: key,
+        hashedOtp,
+        expiresAt: this.getOtpExpiry(),
+      })
+    } else {
+      await this.otpRepository.create({
+        mobileNumber: key,
+        hashedOtp,
+        expiresAt: this.getOtpExpiry(),
+      })
+    }
```

Comment:
wdyt about shifting to `OtpService`? also, side note - i feel like providing a `verifyEmailOtp` + `verifyMobileOtp` (and similarly for the create) method instead of the current approach might be clearer; wdyt?

Reply:
i think similar explanation as above about the intentions of the OtpService

Regarding the methods for email and mobile - I felt enums are a more scalable way. But thinking of it, we only verify email and mobile mostly. There is unlikely to be a 3rd use case. So yea, agree with you that methods might be clearer.

So to clarify, are you suggesting to remove the `OtpType` enum and just directly have separate methods?

Reply:
yeap! it also helps to make the method name more descriptive and keep the method call itself simpler

Reply:
done